File: app/src/main/java/org/breezyweather/search/SearchActivityRepository.java
package org.breezyweather.search;

import android.content.Context;

import androidx.annotation.Nullable;

import java.util.List;

import javax.inject.Inject;

import dagger.hilt.android.qualifiers.ApplicationContext;
import kotlin.Pair;
import org.breezyweather.common.basic.models.Location;
import org.breezyweather.common.basic.models.options.provider.WeatherSource;
import org.breezyweather.common.utils.helpers.AsyncHelper;
import org.breezyweather.main.utils.RequestErrorType;
import org.breezyweather.settings.ConfigStore;
import org.breezyweather.settings.SettingsManager;
import org.breezyweather.weather.WeatherHelper;

public class SearchActivityRepository {

    private final WeatherHelper mWeatherHelper;
    private final ConfigStore mConfig;

    private @Nullable WeatherSource mLastDefaultSourceCache;

    private static final String PREFERENCE_SEARCH_CONFIG = "SEARCH_CONFIG";
    private static final String KEY_DISABLED_SOURCES = "DISABLED_SOURCES";
    private static final String KEY_LAST_DEFAULT_SOURCE = "LAST_DEFAULT_SOURCE";

    private static final String DEFAULT_DISABLED_SOURCES_VALUE = "ENABLE_DEFAULT_SOURCE_ONLY";

    @Inject
    SearchActivityRepository(@ApplicationContext Context context, WeatherHelper weatherHelper) {
        mWeatherHelper = weatherHelper;
        mConfig = new ConfigStore(context, PREFERENCE_SEARCH_CONFIG);

        mLastDefaultSourceCache = null;
    }
    
    public void searchLocationList(Context context, String query, WeatherSource enabledSource,
                                   AsyncHelper.Callback<Pair<List<Location>, RequestErrorType>> callback) {
        mWeatherHelper.requestSearchLocations(context, query, enabledSource, new WeatherHelper.OnRequestLocationListener() {
            @Override
            public void requestLocationSuccess(String query, List<Location> locationList) {
                callback.call(new Pair<>(locationList, null), true);
            }

            @Override
            public void requestLocationFailed(String query, RequestErrorType requestErrorType) {
                callback.call(new Pair<>(null, requestErrorType), true);
            }
        });
    }

    public WeatherSource getValidWeatherSource(Context context) {
        WeatherSource defaultSource = SettingsManager.getInstance(context).getWeatherSource();

        if (defaultSource == mLastDefaultSourceCache) {
            return mLastDefaultSourceCache;
        }

        WeatherSource[] totals = WeatherSource.class.getEnumConstants();
        if (totals == null) {
            return null;
        }

        String lastDefaultSource = mConfig.getString(KEY_LAST_DEFAULT_SOURCE, "");
        mLastDefaultSourceCache = WeatherSource.getInstance(lastDefaultSource);

        String value;
        if (!defaultSource.getId().equals(lastDefaultSource)) {
            // last default source is not equal to current default source which is set by user.

            // we need reset the value.
            mConfig.edit()
                    .putString(KEY_DISABLED_SOURCES, DEFAULT_DISABLED_SOURCES_VALUE)
                    .putString(KEY_LAST_DEFAULT_SOURCE, defaultSource.getId())
                    .apply();
        } else {
            mConfig.edit()
                    .putString(KEY_LAST_DEFAULT_SOURCE, defaultSource.getId())
                    .apply();
        }

        return defaultSource;
    }

    public void setValidWeatherSource(WeatherSource weatherSource) {
        WeatherSource[] totals = WeatherSource.class.getEnumConstants();
        if (totals == null) {
            return;
        }

        StringBuilder b = new StringBuilder();
        for (WeatherSource source : totals) {
            if (!weatherSource.equals(source)) {
                b.append(",").append(source.getId());
            }
        }

        String value = b.length() > 0 ? b.substring(1) : "";
        mConfig.edit().putString(KEY_DISABLED_SOURCES, value).apply();
    }

    public void cancel() {
        mWeatherHelper.cancel();
    }
}

File: app/src/main/java/org/breezyweather/search/SearchActivityRepository.kt
/**
 * This file is part of Breezy Weather.
 *
 * Breezy Weather is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, version 3 of the License.
 *
 * Breezy Weather is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Breezy Weather. If not, see <https://www.gnu.org/licenses/>.
 */

package org.breezyweather.search

import android.content.Context
import breezyweather.domain.location.model.Location
import dagger.hilt.android.qualifiers.ApplicationContext
import io.reactivex.rxjava3.disposables.CompositeDisposable
import io.reactivex.rxjava3.observers.DisposableObserver
import kotlinx.serialization.MissingFieldException
import kotlinx.serialization.SerializationException
import org.breezyweather.BuildConfig
import org.breezyweather.common.exceptions.ApiKeyMissingException
import org.breezyweather.common.exceptions.ApiLimitReachedException
import org.breezyweather.common.exceptions.ApiUnauthorizedException
import org.breezyweather.common.exceptions.InvalidOrIncompleteDataException
import org.breezyweather.common.exceptions.LocationSearchException
import org.breezyweather.common.exceptions.NoNetworkException
import org.breezyweather.common.exceptions.ParsingException
import org.breezyweather.common.extensions.getStringByLocale
import org.breezyweather.common.rxjava.ObserverContainer
import org.breezyweather.common.rxjava.SchedulerTransformer
import org.breezyweather.common.utils.helpers.LogHelper
import org.breezyweather.main.utils.RefreshErrorType
import org.breezyweather.settings.ConfigStore
import org.breezyweather.sources.RefreshHelper
import retrofit2.HttpException
import java.net.SocketTimeoutException
import java.net.UnknownHostException
import java.text.ParseException
import javax.inject.Inject

class SearchActivityRepository @Inject internal constructor(
    @ApplicationContext context: Context,
    private val mRefreshHelper: RefreshHelper,
    private val mCompositeDisposable: CompositeDisposable,
) {
    private val mConfig: ConfigStore = ConfigStore(context, PREFERENCE_SEARCH_CONFIG)

    fun searchLocationList(
        context: Context,
        query: String,
        locationSearchSource: String,
        callback: (t: Pair<List<Location>?, RefreshErrorType?>?, done: Boolean) -> Unit,
    ) {
        mRefreshHelper
            .requestSearchLocations(context, query, locationSearchSource)
            .compose(SchedulerTransformer.create())
            .subscribe(ObserverContainer(mCompositeDisposable, object : DisposableObserver<List<Location>>() {
                override fun onNext(t: List<Location>) {
                    callback(Pair<List<Location>, RefreshErrorType?>(t, null), true)
                }

                override fun onError(e: Throwable) {
                    val refreshErrorType = when (e) {
                        is NoNetworkException -> RefreshErrorType.NETWORK_UNAVAILABLE
                        // Can mean different things but most of the time, it’s a network issue:
                        is UnknownHostException -> RefreshErrorType.NETWORK_UNAVAILABLE
                        is HttpException -> {
                            LogHelper.log(msg = "HttpException ${e.code()}")
                            when (e.code()) {
                                401, 403 -> RefreshErrorType.API_UNAUTHORIZED
                                409, 429 -> RefreshErrorType.API_LIMIT_REACHED
                                in 500..599 -> RefreshErrorType.SERVER_UNAVAILABLE
                                else -> {
                                    e.printStackTrace()
                                    RefreshErrorType.LOCATION_SEARCH_FAILED
                                }
                            }
                        }
                        is ApiLimitReachedException -> RefreshErrorType.API_LIMIT_REACHED
                        is ApiUnauthorizedException -> RefreshErrorType.API_UNAUTHORIZED
                        is SocketTimeoutException -> RefreshErrorType.SERVER_TIMEOUT
                        is ApiKeyMissingException -> RefreshErrorType.API_KEY_REQUIRED_MISSING
                        is MissingFieldException, is SerializationException, is ParsingException,
                        is ParseException -> {
                            e.printStackTrace()
                            RefreshErrorType.PARSING_ERROR
                        }
                        is LocationSearchException -> RefreshErrorType.LOCATION_SEARCH_FAILED
                        is InvalidOrIncompleteDataException -> RefreshErrorType.INVALID_INCOMPLETE_DATA
                        else -> {
                            e.printStackTrace()
                            RefreshErrorType.LOCATION_SEARCH_FAILED
                        }
                    }
                    LogHelper.log(msg = "Refresh error: ${context.getStringByLocale(refreshErrorType.shortMessage)}")
                    callback(
                        Pair<List<Location>, RefreshErrorType?>(emptyList(), refreshErrorType),
                        true
                    )
                }

                override fun onComplete() {
                    // do nothing.
                }
            }))
    }

    var lastSelectedLocationSearchSource: String
        set(value) {
            mConfig.edit().putString(KEY_LAST_DEFAULT_SOURCE, value).apply()
        }
        get() = if (BuildConfig.FLAVOR != "freenet") {
            mConfig.getString(KEY_LAST_DEFAULT_SOURCE, null) ?: BuildConfig.DEFAULT_LOCATION_SEARCH_SOURCE
        } else {
            "openmeteo"
        }

    fun cancel() {
        mCompositeDisposable.clear()
    }

    companion object {
        private const val PREFERENCE_SEARCH_CONFIG = "SEARCH_CONFIG"
        private const val KEY_LAST_DEFAULT_SOURCE = "LAST_DEFAULT_SOURCE"
    }
}
